### DISCUSSION TITLE
[@types/react-native] wrong target in NativeTouchEvent and NativeSyntheticEvent

### DISCUSSION THREAD

**Discussion by eyshella** [DefinitelyTyped/DefinitelyTyped#62550] [Issues with a @types package]
So looks like there is a mismatch between type definitions and real state of affairs. I am using @types/react-native 0.68.2 and react-native 0.68.2, but looks like this issue is actual for the latest versions as well

### **First issue:**
_Steps to reproduce:_ 
1. Use the following template `<View style={{width: 100; height: 100; backgroundColor: 'red'}} onTouchEnd={(event) => console.log(event.target} />`
2. Check output of `onTouchEnd`

_Expected result:_
Type definitions says that `event.target` is a number, so we expect number here.
_Actual result:_
However, it is an object contains View reference.

### **Second issue:**
_Steps to reproduce:_ 
1. Use the following template `<View style={{width: 100; height: 100; backgroundColor: 'red'}} onTouchEnd={(event) => console.log(event.nativeEvent.target} />`
2. Check output of `onTouchEnd`

_Expected result:_
Type definitions says that `event.nativeEvent.target` is a string, so we expect string here.
_Actual result:_
However, it is a number. React-native documentation says that it is something like node id.

[a comment by an automated account is not shown]

**Comment by lunaleaps** [comment]
Regarding first issue:
Yes it seems the TS type doesn't align with Flow's. Would you like to submit a PR to update it?

Regarding second issue: 
I think this is because we use React's types 
```
export interface NativeSyntheticEvent<T>
  extends React.BaseSyntheticEvent<T, NodeHandle, NodeHandle> {}
```

Is that where the mismatch is?

**Reply by eyshella** [reply]
Thanks for your answer, @lunaleaps !
Sure, will submit a PR :) Looks like the _first_ issue is about `React.BaseSyntheticEvent<T, NodeHandle, NodeHandle>` and the second one is about `NativeTouchEvent` interface where `target: string;` is specified for some reason.

**Reply by lunaleaps** [reply]
Oh yup sorry wrong order.

Regarding the first issue, I think we should be defining NodeHandle as ` number | React.ElementRef<HostComponent<unknown>>` and nullable too? 

* See this definition of NodeHandle: https://github.com/facebook/react-native/blob/main/Libraries/ReactNative/RendererProxy.d.ts#L12
* Flow definition: https://github.com/facebook/react-native/blob/main/Libraries/Types/CoreEventTypes.js#L31
* Probably the type we should use: https://github.com/facebook/react-native/blob/main/Libraries/Types/CoreEventTypes.d.ts#L176

**Reply by lunaleaps** [reply]
Would you also like to submit a PR for that? We're currently in migration for types to appear in react-native repo. Once you have PRs for DefinitelyTyped, we can copy them over to react-native

**Reply by eyshella** [reply]
Hm, so NodeHandle is a node identifier, so sound better to me to keep it as number. 
We may want to use NodeHandle type in NativeTouchEvent as `target: NodeHandle` and provide something like `React.ElementRef<HostComponent<unknown>>` to `React.BaseSyntheticEvent<T,  React.ElementRef<HostComponent<unknown>>,  React.ElementRef<HostComponent<unknown>>>`

However, there is a type `number |  React.ElementRef<HostComponent<unknown>> | null` for `target` and `currentTarget` in `CoreEventTypes.js` file, as you provided.
Isn't it a mistake ? It is a bit confusing that we have both types possible here and it requires more work from developers and so on.

The actual behaviour I see looks the following: 
1. event.nativeEvent.target is a number
2. event.target and event.currentTarget are refs

And this makes sense for me to have different types for `event.target` and `event.nativeEvent.target` to give a developer access to both possibilities.
However, I haven't checked the source code closely yet.

Yeah, definitely want to make a PR, but just want to clarify everything first :)

**Reply by lunaleaps** [reply]
Hmm right so the question is why is `event.target` typed as potentially a number? Here is where React creates SyntheticEvents: 
* https://github.com/facebook/react/blob/05c283c3c31184d68c6a54dfd6a044790b89a08a/packages/react-native-renderer/src/ReactNativeEventEmitter.js#L126
* That eventually calls this plugin for the [legacy bridge](https://github.com/facebook/react/blob/b1f34aa307b004681b3d4c75446d354cda6368cb/packages/react-native-renderer/src/ReactNativeBridgeEventPlugin.js#L203) and [this for Fabric](https://github.com/facebook/react/blob/aed33a49ccac5f42ea1bde6e97f2f470edcd6574/packages/react-native-renderer/src/ReactFabricEventEmitter.js#L68).

Also not sure, asking around

**Reply by elicwhite** [reply]
I did some spelunking (from code I wrote two years ago 😅). I think `event.target` is now only a HostComponent. I'm not sure why `event.nativeEvent.target` would still be a number (reactTag) though, that feels like a bug. IIRC, the goal was to remove reactTags from everything, and I probably overlooked changing that at the same time. It does seem like it is *currently* a number though as you identified, so the types should reflect that.

**Reply by eyshella** [reply]
Thanks guys ! I will prepare a PR today's evening.

**Reply by eyshella** [reply]
The [PR](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/62604) is here
